### PR TITLE
Overworld

### PR DESCRIPTION
**Pull request by PikalaxALT**

### CODE REVIEW

File: include/global.fieldmap.h [hunk 1]
```diff
@@ -185,22 +185,31 @@ struct MapConnections
...
+    /* 0x08 */ const u8 *mapScripts;
+    /* 0x0C */ const struct MapConnections *connections;
     /* 0x10 */ u16 music;
     /* 0x12 */ u16 mapLayoutId;
     /* 0x14 */ u8 regionMapSectionId;
     /* 0x15 */ u8 cave;
     /* 0x16 */ u8 weather;
     /* 0x17 */ u8 mapType;
-    /* 0x18 */ u8 filler_18;
+    /* 0x18 */ u8 bikingAllowed;
```
Comment (jiangzhengwenjz) [inline include/global.fieldmap.h]: `bool8`

File: src/script.c [hunk 2]
```diff
@@ -447,7 +447,7 @@ bool8 TryRunOnFrameMapScript(void)
     return 1;
 }
 
-void mapheader_run_first_tag4_script_list_match(void)
+void TryRunOnWarpIntoMapScript(void)
 {
     u8 *ptr = mapheader_get_first_match_from_tagged_ptr_list(4);
     if (ptr)
```
Comment (jiangzhengwenjz) [inline src/script.c]: `if (ptr != NULL)`

File: include/global.fieldmap.h [hunk 3]
```diff
@@ -185,22 +185,31 @@ struct MapConnections
...
 };
 
+// Flags for gMapHeader.flags, as defined in the map_header_flags macro
+#define MAP_ALLOW_BIKE         (1 << 0)
+#define MAP_ALLOW_ESCAPE_ROPE  (1 << 1)
+#define MAP_ALLOW_RUN          (1 << 2)
+#define MAP_SHOW_MAP_NAME      (1 << 2)
+#define UNUSED_MAP_FLAGS       (1 << 3 | 1 << 4 | 1 << 5 | 1 << 6 | 1 << 7)
+
+#define SHOW_MAP_NAME_ENABLED  ((gMapHeader.escapeRope & (MAP_SHOW_MAP_NAME | UNUSED_MAP_FLAGS)) == MAP_SHOW_MAP_NAME)
```
Comment (jiangzhengwenjz) [inline include/global.fieldmap.h]: so `escapeRope` shares the constants with `flags`?

File: src/overworld.c [hunk 4]
```diff
@@ -0,0 +1,3553 @@
...
+{
+    u8 transitionFlags;
+    u8 direction;
+    bool8 unk2;
+};
+
+struct TradeRoomPlayer
+{
+    u8 playerId;
+    u8 isLocalPlayer;
```
Comment (jiangzhengwenjz) [inline src/overworld.c]: `bool8`

File: src/overworld.c [hunk 5]
```diff
@@ -0,0 +1,3553 @@
...
+    u8 c;
+    u8 facing;
+    struct MapPosition pos;
+    u16 field_C;
+};
+
+EWRAM_DATA struct WarpData gLastUsedWarp = {};
+EWRAM_DATA struct WarpData sWarpDestination = {};
```
Comment (jiangzhengwenjz) [inline src/overworld.c]: `static`

File: src/overworld.c [hunk 6]
```diff
@@ -0,0 +1,3553 @@
...
+static EWRAM_DATA struct WarpData gFixedDiveWarp = {};
+static EWRAM_DATA struct WarpData gFixedHoleWarp = {};
+
+// File boundary perhaps?
+static EWRAM_DATA struct InitialPlayerAvatarState gInitialPlayerAvatarState = {};
```
Comment (jiangzhengwenjz) [inline src/overworld.c]: `s` prefix

File: src/overworld.c [hunk 7]
```diff
@@ -0,0 +1,3553 @@
...
+
+static void sub_8057100(void)
+{
+    sub_805EDF0(0, 0);
+    RunOnReturnToFieldMapScript();
+}
+
+static void sub_8057114(void)
+{
+    gObjectEvents[gPlayerAvatar.objectEventId].trackedByCamera = 1;
```
Comment (jiangzhengwenjz) [inline src/overworld.c]: `TRUE`

File: src/overworld.c [hunk 8]
```diff
@@ -0,0 +1,3553 @@
...
+    {
+        CleanupOverworldWindowsAndTilemaps();
+        SetMainCallback2(CB2_LoadMap);
+        DestroyTask(taskId);
+    }
+}
+
+// Link related
+
+static u8 (*const gLinkPlayerMovementModes[])(struct LinkPlayerObjectEvent *, struct ObjectEvent *, u8) = {
```
Comment (jiangzhengwenjz) [inline src/overworld.c]: `sLinkPlayerMovementModes`

File: src/overworld.c [hunk 9]
```diff
@@ -0,0 +1,3553 @@
...
+    MovementEventModeCB_Normal, // MOVEMENT_MODE_FREE
+    MovementEventModeCB_Ignored, // MOVEMENT_MODE_FROZEN
+    MovementEventModeCB_Normal_2, // MOVEMENT_MODE_SCRIPTED
+};
+
+// These handlers return TRUE if the movement was scripted and successful, and FALSE otherwise.
+static bool8 (*const gLinkPlayerFacingHandlers[])(struct LinkPlayerObjectEvent *, struct ObjectEvent *, u8) = {
```
Comment (jiangzhengwenjz) [inline src/overworld.c]: `sLinkPlayerFacingHandlers`

File: src/overworld.c [hunk 10]
```diff
@@ -0,0 +1,3553 @@
...
+    FacingHandler_DoNothing,
+    FacingHandler_DoNothing,
+    FacingHandler_ForcedFacingChange,
+    FacingHandler_ForcedFacingChange,
+    FacingHandler_ForcedFacingChange,
+    FacingHandler_ForcedFacingChange,
+};
+
+// These handlers are run after an attempted movement.
+static void (*const gMovementStatusHandler[])(struct LinkPlayerObjectEvent *, struct ObjectEvent *) = {
```
Comment (jiangzhengwenjz) [inline src/overworld.c]: `sMovementStatusHandler`

File: src/overworld.c [hunk 11]
```diff
@@ -0,0 +1,3553 @@
...
+    SetKeyInterceptCallback(KeyInterCB_SendNothing_2);
+    return 0;
+}
+
+static void LoadTradeRoomPlayer(s32 linkPlayerId, s32 myPlayerId, struct TradeRoomPlayer *trainer)
+{
+    s16 x, y;
+
+    trainer->playerId = linkPlayerId;
+    trainer->isLocalPlayer = (linkPlayerId == myPlayerId) ? 1 : 0;
```
Comment (jiangzhengwenjz) [inline src/overworld.c]: `TRUE` : `FALSE`

File: src/overworld.c [hunk 12]
```diff
@@ -0,0 +1,3553 @@
...
+static void SpawnLinkPlayerObjectEvent(u8 linkPlayerId, s16 x, s16 y, u8 a4)
+{
+    u8 objEventId = GetFirstInactiveObjectEventId();
+    struct LinkPlayerObjectEvent *linkPlayerObjEvent = &gLinkPlayerObjectEvents[linkPlayerId];
+    struct ObjectEvent *objEvent = &gObjectEvents[objEventId];
+
+    ZeroLinkPlayerObjectEvent(linkPlayerObjEvent);
+    ZeroObjectEvent(objEvent);
+
+    linkPlayerObjEvent->active = 1;
```
Comment (jiangzhengwenjz) [inline src/overworld.c]: `TRUE`

File: src/overworld.c [hunk 13]
```diff
@@ -0,0 +1,3553 @@
...
+    linkPlayerObjEvent->linkPlayerId = linkPlayerId;
+    linkPlayerObjEvent->objEventId = objEventId;
+    linkPlayerObjEvent->movementMode = MOVEMENT_MODE_FREE;
+
+    objEvent->active = 1;
```
Comment (jiangzhengwenjz) [inline src/overworld.c]: `TRUE`

File: src/overworld.c [hunk 14]
```diff
@@ -0,0 +1,3553 @@
...
+
+static void sub_80584B8(u8 linkPlayerId)
+{
+    struct LinkPlayerObjectEvent *linkPlayerObjEvent = &gLinkPlayerObjectEvents[linkPlayerId];
+    u8 objEventId = linkPlayerObjEvent->objEventId;
+    struct ObjectEvent *objEvent = &gObjectEvents[objEventId];
+    if (objEvent->spriteId != MAX_SPRITES)
+        DestroySprite(&gSprites[objEvent->spriteId]);
+    linkPlayerObjEvent->active = 0;
+    objEvent->active = 0;
```
Comment (jiangzhengwenjz) [inline src/overworld.c]: `FALSE`

File: src/overworld.c [hunk 15]
```diff
@@ -0,0 +1,3553 @@
...
+{
+    struct LinkPlayerObjectEvent *linkPlayerObjEvent = &gLinkPlayerObjectEvents[linkPlayerId];
+    u8 objEventId = linkPlayerObjEvent->objEventId;
+    struct ObjectEvent *objEvent = &gObjectEvents[objEventId];
+
+    if (linkPlayerObjEvent->active)
+    {
+        if (facing > FACING_FORCED_RIGHT)
+        {
+            objEvent->triggerGroundEffectsOnMove = 1;
```
Comment (jiangzhengwenjz) [inline src/overworld.c]: `TRUE`

File: src/overworld.c [hunk 16]
```diff
@@ -0,0 +1,3553 @@
...
+    case FACING_FORCED_LEFT:
+        return DIR_WEST;
+    case FACING_RIGHT:
+    case FACING_FORCED_RIGHT:
+        return DIR_EAST;
+    }
+    return oldFacing;
+}
+
+static u8 LinkPlayerDetectCollision(u8 selfObjEventId, u8 a2, s16 x, s16 y)
```
Comment (jiangzhengwenjz) [inline src/overworld.c]: return value is `bool8`

File: src/overworld.c [hunk 17]
```diff
@@ -0,0 +1,3553 @@
...
+        }
+        else
+        {
+            objEvent->spriteId = AddPseudoObjectEvent(GetRSAvatarGraphicsIdByGender(objEvent->singleMovementActive), SpriteCB_LinkPlayer, 0, 0, 0);
+        }
+
+        sprite = &gSprites[objEvent->spriteId];
+        sprite->coordOffsetEnabled = TRUE;
+        sprite->data[0] = linkPlayerId;
+        objEvent->triggerGroundEffectsOnMove = 0;
```
Comment (jiangzhengwenjz) [inline src/overworld.c]: `FALSE`

File: include/global.fieldmap.h (hunk 3 above)
Comment (PikalaxALT) [reply include/global.fieldmap.h]: Field is misnamed, fixing for next commit.

**Pull request merged**